### PR TITLE
Fix cbars for different norms

### LINKED ISSUE #16280: SymLogNorm colorbar incorrect on master
### Bug report

**Bug summary**

The colormap/tick locations within the colorbar are not being updated properly with a SymLogNorm on master. I think this is different than #16269 because it is actually creating the colorbar, just a wrong one.

Compare the previous correct version:
https://matplotlib.org/3.1.0/gallery/userdemo/colormap_normalizations_symlognorm.html

with a fresh checkout of master:

test

Difference in the colorbar of the top axis.

### PR DESCRIPTION
**Pull request by dstansby**
Fixes #16280, and also fixes colorbars with a few other norms.

I have also added an image test. I can't think of a way to test this without an image, but suggestions welcome. Before making this change, the test image with different norms looks like this on current master:

colorbar_norms_before

Whereas after the fix it is:

colorbar_norms

cc @jklymak

### CODE REVIEW

**Comment by jklymak** [comment]
Obviously the test is a good idea!  Not sure what to do about the failure though ;-)

File: lib/matplotlib/tests/test_colorbar.py [hunk 1]
```diff
@@ -570,3 +570,17 @@ def test_colorbar_label():
 
     cbar3 = fig.colorbar(im, orientation='horizontal', label='horizontal cbar')
     assert cbar3.ax.get_xlabel() == 'horizontal cbar'
+
+@image_comparison(
+    baseline_images=['colorbar_norms'], extensions=['png'])
```
Comment (anntzer) [inline lib/matplotlib/tests/test_colorbar.py]: Image test looks reasonable (no need to *overthink* into avoiding them either); perhaps you can use style="default"?  (ticks inwards look weird to me now :-), and then you don't need to specify viridis below explicitly)

**Comment by greglucas** [comment]
Thanks for looking at this so quickly, @dstansby! To get around the images I thought you might be able to look at coordinates from a transform within the axes and compare them to the norm somehow (make a 0-1 axis and then `tr = cbar.ax.transData + fig.transFigure.inverted()` or something similar to back out where the tick(x) is versus the norm(x), but I think the issue is that these norms that fail don't have an inverse which defeats that route. Maybe you can think of a better way of doing that though to avoid the image test if you're inclined. But, I also think the image fine.

**Comment by anntzer** [comment]
test failure seems real

**Comment by dstansby** [comment]
Yep, they are. Not sure I have time in the next couple of days to fix, so if anyone else fancies investigating, fixing and pushing to my branch before then feel free.

**Comment by jklymak** [comment]
I pushed to your branch but ci doesn’t seem to like it.  Can your force push the change?

**Comment by jklymak** [comment]
I figured out the force push

**Comment by jklymak** [comment]
This is a situation where the Norm not having an accompanying scale is what is causing all the problems.  The current fix is fine, but it basically special cases `LogNorm` and `Normalize`, and everything else is handled on a fake linear scale.  

For `SymLogNorm` and `LogitNorm` there are scales, but I don't see one for `PowerNorm`.  

My suggestion would be for a bit of re-architecting here, and make each norm have a scale property that colorbar could consult rather than us having to guess the scale.  That would also trigger folks who write Norms to create a corresponding scale so the colorbar can work with the a real y scale rather than a faked one.

**Comment by jklymak** [comment]
I'm OK with this as is for a temporary fix - it passes CI except for a Pandas deprecation warning, which isn't this PR.

**Comment by jklymak** [comment]
I also don't think the symmetric log plot is correct above - the linear scale is supposed to be one decade of the logarithmic scale, and its clearly not.  It would be interesting to see if it ever was, but I doubt it.  This will take some sitting down to figure out and may take me a few weeks as I have other deadlines.

**Comment by dstansby** [comment]
> I also don't think the symmetric log plot is correct above - the linear scale is supposed to be one decade of the logarithmic scale, and its clearly not. It would be interesting to see if it ever was, but I doubt it. This will take some sitting down to figure out and may take me a few weeks as I have other deadlines.

The SymLog code has been broken forever I think, and I have been working on it on and off recently. Hopefully I'll have something to show for it soon, thanks for opening the issue.

**Comment by jklymak** [comment]
Given that @anntzer has a factory for turning scales into norms, that would really make our lives a *lot* easier as we'd only need to make sure the Transform is correct, rather than rewriting one in the Norm.  I think if we have a colornorm it should have a corresponding scale for the colorbar, so enforcing some regularity would be good.

**Comment by jklymak** [comment]
Sorry I botched this up.  I'll push this as a new PR with my apologies.

**Pull request closed without merging**